From d1703ffd2b4cd3ffa163e5cf919af11408beb3a9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Tue, 23 Jan 2024 17:20:20 +0100
Subject: [PATCH] menu: move lock and logout menu items below the user menu

This makes the main menu less cluttered and consistent with the user menu applet.
---
 modules/menu/gp-menu-button-applet.c | 40 +++++++++++++++++++++++-----
 modules/menu/gp-menu-module.c        | 21 +++++++++++++--
 2 files changed, 53 insertions(+), 8 deletions(-)

diff --git a/modules/menu/gp-menu-button-applet.c b/modules/menu/gp-menu-button-applet.c
index b864b9c8a..0f975c26a 100644
--- a/modules/menu/gp-menu-button-applet.c
+++ b/modules/menu/gp-menu-button-applet.c
@@ -220,16 +220,26 @@ append_places_item (GpMenuButtonApplet *menu_button,
                           G_BINDING_SYNC_CREATE);
 }
 
+static void
+append_lock_logout (GtkMenu                   *menu,
+                    GpMenuButtonAppletPrivate *priv)
+{
+  gp_lock_logout_append_to_menu (priv->lock_logout, menu);
+}
+
 static void
 append_user_item (GpMenuButtonApplet *menu_button,
                   GtkMenu            *menu)
 {
+  GpMenuButtonAppletPrivate *priv;
   guint icon_size;
   GtkWidget *icon;
   gchar *user_name;
   GtkWidget *item;
   GtkWidget *user_menu;
 
+  priv = gp_menu_button_applet_get_instance_private (menu_button);
+
   icon_size = gp_applet_get_menu_icon_size (GP_APPLET (menu_button));
   icon = gtk_image_new_from_icon_name ("computer", GTK_ICON_SIZE_MENU);
   gtk_image_set_pixel_size (GTK_IMAGE (icon), icon_size);
@@ -266,21 +276,39 @@ append_user_item (GpMenuButtonApplet *menu_button,
   g_object_bind_property (user_menu, "empty", item, "visible",
                           G_BINDING_DEFAULT | G_BINDING_SYNC_CREATE |
                           G_BINDING_INVERT_BOOLEAN);
+
+  priv->lock_logout = gp_lock_logout_new ();
+
+  g_object_bind_property (menu_button, "enable-tooltips",
+                          priv->lock_logout, "enable-tooltips",
+                          G_BINDING_DEFAULT |
+                          G_BINDING_SYNC_CREATE);
+
+  g_object_bind_property (menu_button, "locked-down",
+                          priv->lock_logout, "locked-down",
+                          G_BINDING_DEFAULT |
+                          G_BINDING_SYNC_CREATE);
+
+  g_object_bind_property (menu_button, "menu-icon-size",
+                          priv->lock_logout, "menu-icon-size",
+                          G_BINDING_DEFAULT |
+                          G_BINDING_SYNC_CREATE);
+
+  g_signal_connect_swapped (priv->lock_logout, "changed",
+                            G_CALLBACK (gp_user_menu_reload), user_menu);
+
+  gp_user_menu_set_append_func (GP_USER_MENU (user_menu),
+                                (GpAppendMenuItemsFunc) append_lock_logout,
+                                priv);
 }
 
 static void
 append_menu_items_cb (GtkMenu            *menu,
                       GpMenuButtonApplet *menu_button)
 {
-  GpMenuButtonAppletPrivate *priv;
-
-  priv = gp_menu_button_applet_get_instance_private (menu_button);
-
   append_separator_if_needed (menu);
   append_places_item (menu_button, menu);
   append_user_item (menu_button, menu);
-
-  gp_lock_logout_append_to_menu (priv->lock_logout, menu);
 }
 
 static gboolean
diff --git a/modules/menu/gp-menu-module.c b/modules/menu/gp-menu-module.c
index 966bd5782..b9da2e20e 100644
--- a/modules/menu/gp-menu-module.c
+++ b/modules/menu/gp-menu-module.c
@@ -476,6 +476,13 @@ append_places_item (StandaloneMenuData *data,
                 NULL);
 }
 
+static void
+append_lock_logout (GtkMenu            *menu,
+                    StandaloneMenuData *data)
+{
+  gp_lock_logout_append_to_menu (data->lock_logout, menu);
+}
+
 static void
 append_user_item (StandaloneMenuData *data,
                   GtkMenu            *menu)
@@ -509,6 +516,18 @@ append_user_item (StandaloneMenuData *data,
   g_object_bind_property (user_menu, "empty", item, "visible",
                           G_BINDING_DEFAULT | G_BINDING_SYNC_CREATE |
                           G_BINDING_INVERT_BOOLEAN);
+
+  data->lock_logout = gp_lock_logout_new ();
+
+  g_object_set (data->lock_logout,
+                "enable-tooltips", data->enable_tooltips,
+                "locked-down", data->locked_down,
+                "menu-icon-size", data->menu_icon_size,
+                NULL);
+
+  gp_user_menu_set_append_func (GP_USER_MENU (user_menu),
+                                (GpAppendMenuItemsFunc) append_lock_logout,
+                                data);
 }
 
 static void
@@ -518,8 +537,6 @@ append_menu_items_cb (GtkMenu            *menu,
   append_separator_if_needed (GTK_MENU (menu));
   append_places_item (data, menu);
   append_user_item (data, menu);
-
-  gp_lock_logout_append_to_menu (data->lock_logout, GTK_MENU (menu));
 }
 
 static GtkWidget *
-- 
2.43.0

